plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts

```diff
@@ -0,0 +1,363 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { PluginTaskScheduler, TaskRunner } from '@backstage/backend-tasks';
+import { Config } from '@backstage/config';
+import { GitLabIntegration, ScmIntegrations } from '@backstage/integration';
+import {
+  EntityProvider,
+  EntityProviderConnection,
+} from '@backstage/plugin-catalog-backend';
+import * as uuid from 'uuid';
+import { Logger } from 'winston';
+import {
+  GitLabClient,
+  GitlabProviderConfig,
+  paginated,
+  readGitlabConfigs,
+} from '../lib';
+import { GitLabGroup, GitLabUser } from '../lib/types';
+import {
+  ANNOTATION_LOCATION,
+  ANNOTATION_ORIGIN_LOCATION,
+  Entity,
+  UserEntity,
+  GroupEntity,
+} from '@backstage/catalog-model';
+import { merge } from 'lodash';
+
+type Result = {
+  scanned: number;
+  matches: GitLabUser[];
+};
+
+type GroupResult = {
+  scanned: number;
+  matches: GitLabGroup[];
+};
+
+/**
+ * Discovers entity definition files in the groups of a Gitlab instance.
+ * @public
+ */
+export class GitlabOrgDiscoveryEntityProvider implements EntityProvider {
+  private readonly config: GitlabProviderConfig;
+  private readonly integration: GitLabIntegration;
+  private readonly logger: Logger;
+  private readonly scheduleFn: () => Promise<void>;
+  private connection?: EntityProviderConnection;
+
+  static fromConfig(
+    config: Config,
+    options: {
+      logger: Logger;
+      schedule?: TaskRunner;
+      scheduler?: PluginTaskScheduler;
+    },
+  ): GitlabOrgDiscoveryEntityProvider[] {
+    if (!options.schedule && !options.scheduler) {
+      throw new Error('Either schedule or scheduler must be provided.');
+    }
+
+    const providerConfigs = readGitlabConfigs(config);
+    const integrations = ScmIntegrations.fromConfig(config).gitlab;
+    const providers: GitlabOrgDiscoveryEntityProvider[] = [];
+
+    providerConfigs.forEach(providerConfig => {
+      const integration = integrations.byHost(providerConfig.host);
+      if (!integration) {
+        throw new Error(
+          `No gitlab integration found that matches host ${providerConfig.host}`,
+        );
+      }
+
+      if (!options.schedule && !providerConfig.schedule) {
+        throw new Error(
+          `No schedule provided neither via code nor config for GitlabOrgDiscoveryEntityProvider:${providerConfig.id}.`,
+        );
+      }
+
+      if (!providerConfig.orgEnabled) {
```

Comment:
Shouldn't this be moved to be handled before the `.schedule` check?

Reply:
I've moved it. It doesn't really matter, cause the other gitlab integration check is enabled by default when configured anyway and checks that too.